env/file_system_tracer.cc

```diff
@@ -0,0 +1,253 @@
+  io_tracer_->WriteIOOp(io_record);
+  return s;
+}
+
+IOStatus FileSystemTracingWrapper::DeleteDir(const std::string& dirname,
+                                             const IOOptions& options,
+                                             IODebugContext* dbg) {
+  IOStatus s = target()->DeleteDir(dirname, options, dbg);
+  IOTraceRecord io_record(Timestamp(), TraceType::kIOFileName, "DeleteDir",
+                          s.ToString(), dirname);
+  io_tracer_->WriteIOOp(io_record);
+  return s;
+}
+
+IOStatus FileSystemTracingWrapper::GetFileSize(const std::string& fname,
+                                               const IOOptions& options,
+                                               uint64_t* file_size,
+                                               IODebugContext* dbg) {
+  IOStatus s = target()->GetFileSize(fname, options, file_size, dbg);
+  IOTraceRecord io_record(Timestamp(), TraceType::kIOFileNameAndFileSize,
+                          "GetFileSize", s.ToString(), fname, *file_size);
+  io_tracer_->WriteIOOp(io_record);
+  return s;
+}
+
+IOStatus FSSequentialFileTracingWrapper::Read(size_t n,
+                                              const IOOptions& options,
+                                              Slice* result, char* scratch,
+                                              IODebugContext* dbg) {
+  IOStatus s = target()->Read(n, options, result, scratch, dbg);
+  IOTraceRecord io_record(Timestamp(), TraceType::kIOLen, "Read", s.ToString(),
+                          0 /* offset */, result->size());
+  io_tracer_->WriteIOOp(io_record);
+  return s;
+}
+
+IOStatus FSSequentialFileTracingWrapper::InvalidateCache(size_t offset,
+                                                         size_t length) {
+  IOStatus s = target()->InvalidateCache(offset, length);
+  IOTraceRecord io_record(Timestamp(), TraceType::kIOOffsetAndLen,
+                          "InvalidateCache", s.ToString(), offset, length);
+  io_tracer_->WriteIOOp(io_record);
+  return s;
+}
+
+IOStatus FSSequentialFileTracingWrapper::PositionedRead(
+    uint64_t offset, size_t n, const IOOptions& options, Slice* result,
+    char* scratch, IODebugContext* dbg) {
+  IOStatus s =
+      target()->PositionedRead(offset, n, options, result, scratch, dbg);
+  IOTraceRecord io_record(Timestamp(), TraceType::kIOOffsetAndLen,
+                          "PositionedRead", s.ToString(), offset,
+                          result->size());
+  io_tracer_->WriteIOOp(io_record);
+  return s;
+}
+
+IOStatus FSRandomAccessFileTracingWrapper::Read(uint64_t offset, size_t n,
+                                                const IOOptions& options,
+                                                Slice* result, char* scratch,
+                                                IODebugContext* dbg) const {
+  IOStatus s = target()->Read(offset, n, options, result, scratch, dbg);
+  IOTraceRecord io_record(Timestamp(), TraceType::kIOOffsetAndLen, "Read",
```

Comment:
When the API like Read, Append, are we able to know which file is accessed by this request?

Reply:
For some cases we can. I am not sure about all because its being used at multiple instances. My next PR that i am working on is to cache FileSystemPtr and other Ptrs instead of storing FileSystem and other pointers. So while making those changes i will have better idea where can I pass the file name as part of IODebugContext. I will add the filename in the next PR.

Reply:
I see, that will be great.